drivers: video: sw_pipeline: glue lib/pixel and drivers/video

josuah opened the pull request:
Dependencies:

- https://github.com/zephyrproject-rtos/zephyr/pull/86671
- https://github.com/zephyrproject-rtos/zephyr/pull/76798

Sibling:

- https://github.com/zephyrproject-rtos/zephyr/pull/87007

This allows some very short applications be written to see a video stream from their host as a webcam

```c
#define WIDTH_IN 64
#define HEIGHT_IN 64

#define WIDTH_OUT 128
#define HEIGHT_OUT 32

void app_tune_rgb24line(const uint8_t *rgb24in, uint8_t *rgb24out, uint16_t width)
{
        for (size_t w = 0; w < width; w++) {
                rgb24out[w * 3 + 0] = CLAMP((-64 + rgb24in[w * 3 + 0]) * 2, 0x00, 0xff);
                rgb24out[w * 3 + 1] = CLAMP((-64 + rgb24in[w * 3 + 1]) * 1, 0x00, 0xff);
                rgb24out[w * 3 + 2] = CLAMP((-64 + rgb24in[w * 3 + 2]) * 2, 0x00, 0xff);
        }
}

PIXEL_RGB565LESTREAM_TO_RGB24STREAM(step_rgb565_to_rgb24, WIDTH_IN, HEIGHT_IN);
PIXEL_SUBSAMPLE_RGB24STREAM(step_resize_rgb24, WIDTH_IN, HEIGHT_IN);
PIXEL_RGB24LINE_DEFINE(step_tune_rgb24, app_tune_rgb24line, WIDTH_OUT, HEIGHT_OUT);
PIXEL_RGB24STREAM_TO_YUYVSTREAM_BT709(step_rgb24_to_yuyv, WIDTH_OUT, HEIGHT_OUT);

int app_init_pipeline(void)
{
        const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(video_sw_pipeline));
        struct pixel_stream *strm;

        strm = pixel_stream(&step_rgb565_to_rgb24, &step_resize_rgb24, &step_tune_rgb24,
                            &step_rgb24_to_yuyv, NULL);
        video_sw_pipeline_set_pipeline(dev, strm, VIDEO_PIX_FMT_RGB565, WIDTH_IN, HEIGHT_IN,
                                       VIDEO_PIX_FMT_YUYV, WIDTH_OUT, HEIGHT_OUT);
        return 0;
}

SYS_INIT(app_init_pipeline, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEVICE);
```

Input from the webcam (after commenting out `&step_resize_rgb24, &step_tune_rgb24,` then `WIDTH/HEIGHT_OUT=64`):
![video4-before](https://github.com/user-attachments/assets/4824b5af-2161-4683-b6da-76dc467ffe98)

Input from the webcam (after commenting out `&step_tune_rgb24`):
![video4-middle](https://github.com/user-attachments/assets/343e87ff-6847-4ed4-98e4-e23e10543b5f)

Output from the webcam (nothing commented):
![video4-after](https://github.com/user-attachments/assets/ab5e6aff-9dc5-4c37-8dba-b8526765bded)

josuah committed:
samples: usb: add new UVC sample

Following the addition of USB Video Class, this adds a sample that makes
use of the &zephyr,camera chosen node of any board to stream the video
source to the host. A fallback video-emul.overlay is provided for test
and debugging purpose for devices without a camera.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
usb: device_next: new USB Video Class implementation

Introduce a new USB Video Class (UVC) implementation from scratch.
It exposes a native Zephyr Video driver interface, allowing to call the
video_enqueue()/video_dequeue() interface. It will query the attached
video device to learn about the pipeline capabilities, and use this to
configure the USB descriptors. At runtime, this UVC implementation will
send this device all the control requests, which it can then dispatch to
the rest of the pipeline. The application can poll the format currently
selected by the host, but will not be alerted when the host configures
a new format, as there is no video.h API for it yet.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
boards: nicla_vision: add usbd test feature

The Arduino Nicla Vision board supports the new device_next USB stack
and can be used for testing USB Device features such as UVC.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
lib: pixel: Introduce a pixel and image processing library

The "pixel" library aims facilitating the implementation of all image
processing tasks, such as pixel conversion, with a focus on low-resource
environments. The processing functions scale down to per-pixel "kernels"
to line-based conversion to full streams of several frames.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: lib: pixel: show usage of the pixel library

The newly introduced lib/pixel features utilities that help composing
video pipelines together for the purpose of stream processing, as well
as debug utilities.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: sw_pipeline: perform custom pixel stream processing

Introduce a connector between the lib/pixel API to the drivers/video API.
An function loads a lib/pixel stream into this new driver, then the
driver will load the frames in and out of this stream.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
drivers: video: sw_generator: convert to use the devicetree

Make the video software generator a devicetree node, which allows enabling
several instances, and select it as chosen { zephyr,camera = &... }; node.
It can be enabled via a `video-sw-generator` snippet.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: convert to new LOG_MODULE_REGISTER() syntax

Convert the log module declarationto the new syntax introduced in
5e34681 proposing a more compact syntax.
This fixes the log level not being updated in the "capture" sample.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: tcpserversink: add _dev suffix to variable

Variables that holds pointers to 'const struct device' typically end-up
with '_dev'. The tcpserversink video sample did not have it. Add it.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
doc: build: dts: api: document the zephyr,camera chosen node

The zephyr,camera was used in several samples, but was not documented.
Add the missing documentation entry for it in the list.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah committed:
samples: drivers: video: fix board definitions

By having an "auto-fallback" mechanism, the VIDEO_SW_GENERATOR did shadow
the fact that "platform:mimxrt1064_evk:SHIELD=dvp_fpc24_mt9m114" was
missing the full specification, and therefore was not matched at all:
the shield was not selected, but the CI still worked.

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
The logs give performance statistics for each block of the pipeline:

```
[00:00:49.621,000] <inf> pixel_stream:   638 us on [step_rgb565_to_rgb24 (pixel_rgb565lestream_to_rgb24stream) ((64))x((64))]
[00:00:49.621,000] <inf> pixel_stream:  1316 us on [step_rgb24_to_yuyv (pixel_rgb24stream_to_yuyvstream_bt709) ((64))x((64))]
[00:00:49.621,000] <inf> pixel_stream:     0 us on [export video_sw_pipeline]
```

josuah commented:
Nothing specific to UVC though: network capture or a display also work for instance.

This is a video device for connecting to any input/output and do some processing in-between.

josuah commented:
Converted to draft until the dependencies are merged :)

josuah commented on include/zephyr/usb/class/usbd_uvc.h:
```diff
@@ -0,0 +1,21 @@
+
+/**
+ * @file
+ * @brief USB Device Firmware Upgrade (DFU) public header
+ *
+ * Header exposes API for registering DFU images.
```
This is from another PR, I will check that it is fixed at the source :grimacing:

josuah commented on include/zephyr/pixel/resize.h:
```diff
@@ -0,0 +1,78 @@
+ * @param src_width Width of the input in number of pixels.
+ * @param src_height Height of the input in number of pixels.
+ * @param dst_buf Output buffer in which the stretched/compressed is stored.
+ * @param dst_width Width of the outputin number of pixels.
+ * @param dst_height Height of the outputin number of pixels.
```
This is from another PR, I will check that it is fixed at the source :grimacing:

josuah commented on include/zephyr/pixel/kernel.h:
```diff
@@ -0,0 +1,135 @@
+#include <zephyr/pixel/stream.h>
+
+/**
+ * @brief Define an 3x3 identity kernel operation for RGB24 data.
```
This is from another PR, I will check that it is fixed at the source :grimacing:

josuah commented on include/zephyr/pixel/stats.h:
```diff
@@ -0,0 +1,139 @@
+#include <stdint.h>
+
+/**
+ * @brief Collect red, gren, blue channel averages of all pixels in an RGB24 frame.
```
This is from another PR, I will check that it is fixed at the source :grimacing:

nordicjm commented on drivers/video/Kconfig.sw_pipeline:
```diff
@@ -0,0 +1,17 @@
+config VIDEO_SW_PIPELINE_THREAD_PRIORITY
+	int "Video Software Pipeline thread priority"
+	default 2
+	help
```
`depends on VIDEO_SW_PIPELINE`

nordicjm commented on lib/pixel/CMakeLists.txt:
```diff
@@ -0,0 +1,13 @@
+zephyr_library_sources(bayer.c)
+zephyr_library_sources(formats.c)
+zephyr_library_sources(kernel.c)
+zephyr_library_sources(print.c)
```
shouldn't this only be included if the Kconfig is set?

josuah replied:
AFAIU it is already done from the [parent `CMakeLists.txt`]()https://github.com/zephyrproject-rtos/zephyr/pull/88839/files#diff-4803677bd280f4dded52c428e3c8ac6c618508dcabcbe880b68be85897c2ad41:
```
add_subdirectory_ifdef(CONFIG_PIXEL pixel)
```

Let me know if there should also be a `_ifdef` condition on the inside.

josuah replied:
`CONFIG_PIXEL_PRINT_NONE=y` allows to keep all the `pixel_print_` debug statements which help for troubleshooting the tests and useful in samples, but turn them off the rest of the time.
So when it is on, the `pixel.c` functions are still called but do nothing.
Bad idea?

nordicjm commented on samples/drivers/video/sw_pipeline/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+project(usb_video)
+
+include(${ZEPHYR_BASE}/samples/subsys/usb/common/common.cmake)
+target_sources(app PRIVATE src/main.c)
+target_sources(app PRIVATE src/pipeline.c)
```
```suggestion
target_sources(app PRIVATE src/main.c src/pipeline.c)
```

josuah replied:
Good to know about this, I will preserve this style in future PRs.

nordicjm commented on samples/drivers/video/sw_pipeline/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(usb_video)
```
should match name of sample

josuah replied:
I must have forgotten to change it. Adjusted thank you.

josuah committed:
drivers: video: controls: add the BASE and CAMERA controls

Add all the base controls present like they are in Linux into Zephyr,
limited to those that can apply in the current system:
- Buttons are left as integer for now.
- Some description is modified to fit the Zephyr situation.
- For the minimum number of buffer, Zephyr uses a different mechanism.
- No audio support through the video subsystem.
- Homogenize the wording

Signed-off-by: Josuah Demangeon <me@josuah.net>

josuah commented:
Force-push:

- Rebase on latest UVC branch

Note, this might crrash due to a bug in the video-sw-generator fixed in https://github.com/zephyrproject-rtos/zephyr/pull/85968 (not merged yet).

In order to test it, you will need to specify a framerate manually, such as `ffplay -framerate 30 /dev/video4`
